ff/src/fields/models/fp/mod.rs

```diff
@@ -0,0 +1,1015 @@
+    type FftConfig = P;
+
+    #[inline]
+    fn two_adic_root_of_unity() -> Self {
+        P::TWO_ADIC_ROOT_OF_UNITY
+    }
+
+    #[inline]
+    fn large_subgroup_root_of_unity() -> Option<Self> {
+        P::LARGE_SUBGROUP_ROOT_OF_UNITY
+    }
+
+    #[inline]
+    fn multiplicative_generator() -> Self {
+        P::GENERATOR
+    }
+}
+
+impl<P: FpConfig<N>, const N: usize> SquareRootField for Fp<P, N> {
+    #[inline]
+    fn legendre(&self) -> LegendreSymbol {
+        use crate::fields::LegendreSymbol::*;
+
+        // s = self^((MODULUS - 1) // 2)
+        let s = self.pow(Self::MODULUS_MINUS_ONE_DIV_TWO);
+        if s.is_zero() {
+            Zero
+        } else if s.is_one() {
+            QuadraticResidue
+        } else {
+            QuadraticNonResidue
+        }
+    }
+
+    #[inline]
+    fn sqrt(&self) -> Option<Self> {
+        P::square_root(self)
+    }
+
+    fn sqrt_in_place(&mut self) -> Option<&mut Self> {
+        (*self).sqrt().map(|sqrt| {
+            *self = sqrt;
+            self
+        })
+    }
+}
+
+/// Note that this implementation of `Ord` compares field elements viewing
+/// them as integers in the range 0, 1, ..., P::MODULUS - 1. However, other
+/// implementations of `PrimeField` might choose a different ordering, and
+/// as such, users should use this `Ord` for applications where
+/// any ordering suffices (like in a BTreeMap), and not in applications
+/// where a particular ordering is required.
+impl<P: FpConfig<N>, const N: usize> Ord for Fp<P, N> {
+    #[inline(always)]
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.into_bigint().cmp(&other.into_bigint())
+    }
+}
+
+/// Note that this implementation of `PartialOrd` compares field elements viewing
+/// them as integers in the range 0, 1, ..., `P::MODULUS` - 1. However, other
+/// implementations of `PrimeField` might choose a different ordering, and
+/// as such, users should use this `PartialOrd` for applications where
+/// any ordering suffices (like in a BTreeMap), and not in applications
+/// where a particular ordering is required.
+impl<P: FpConfig<N>, const N: usize> PartialOrd for Fp<P, N> {
+    #[inline(always)]
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<P: FpConfig<N>, const N: usize> From<u128> for Fp<P, N> {
+    fn from(other: u128) -> Self {
+        let mut default_int = BigInt::default();
+        if N == 1 {
+            default_int.0[0] = (other % u128::from(P::MODULUS.0[0])) as u64;
+        } else {
+            let upper = (other >> 64) as u64;
+            let lower = ((other << 64) >> 64) as u64;
+            // This is equivalent to the following, but satisfying the compiler:
+            // default_int.0[0] = lower;
+            // default_int.0[1] = upper;
+            let limbs = [lower, upper];
+            for (cur, other) in default_int.0.iter_mut().zip(&limbs) {
+                *cur = *other;
+            }
```

Comment:
Can we remove this hack now?

Reply:
I feel that the commented code is clearly preferred.
`default_int.0[0] = lower;
default_int.0[1] = upper;`

Reply:
```suggestion
            let upper = (other >> 64) as u64;
            let lower = ((other << 64) >> 64) as u64;
            default_int.0[0] = lower;
            default_int.0[1] = upper;
```